---
title: "1_Dada2_pipeline_rats_hiperamm"
output: html_document
---

```{r Dada2 installation}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("dada2")
```

```{r Dada2 run}
library(dada2); packageVersion("dada2")
path <- "/fast/AG_Forslund/Lola/Datos_CIPF_2023/Seq_rats" 
list.files(path)


# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_R"), `[`, 1)

plotQualityProfile(fnFs[1:2])
plotQualityProfile(fnRs[1:2])

# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))


out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(280,265),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE, trimLeft = c(18, 21),
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)


errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

plotErrors(errF, nominalQ=TRUE)

derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
# Name the derep-class objects by the sample names
names(derepFs) <- sample.names
names(derepRs) <- sample.names

dadaFs <- dada(derepFs, err=errF, multithread=TRUE)
dadaRs <- dada(derepRs, err=errR, multithread=TRUE)
dadaFs[[1]]

mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])

seqtab <- makeSequenceTable(mergers)
dim(seqtab)
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
#Reduce noise
#262 302 303 367 386 400 401 402 403 404 405 406 407 408 409 411 412 413 415 417 418 419 420 421 422 424 425 426 427 441 
#  3   3   2   1   2  67 675 351 244 296   5   6  24   1   1   1   1   1   1   1  38 157 175 363 125   1  25 163  19   1 
seqtab2 <- seqtab[,nchar(colnames(seqtab)) %in% seq(400,427)]

seqtab.nochim <- removeBimeraDenovo(seqtab2, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab2)

getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)

colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)

##Create an ASV matrix with ASV as rows and samples as columns
asvmat <- t(seqtab.nochim) #Removing sequence rownames for display only
asv.names <- as.data.frame(rownames(asvmat))
rownames(asv.names) <- paste0("ASV", 1:nrow(asv.names))
rownames(asvmat)<-NULL
rownames(asvmat) <- paste0("ASV", 1:nrow(asvmat))
#colnames(asvmat) <- paste0("Sample", 1:ncol(asvmat)) #keep original sample ID
head(asvmat)
library(repmod)
make_csv_table(asvmat, "/fast/AG_Forslund/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/Tables/otu_table_names", info = "otu_table")
make_csv_table(seqtab.nochim, "/fast/AG_Forslund/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/Tables/otu_table_with_seqs", info = "otu_table_with_seqs")
write.csv(asvmat, "/fast/AG_Forslund/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/Tables/otu_table.csv", row.names = F, col.names = F)

##Get count of ASVs detected by sample
asv.sample<- as.data.frame(asvmat)
test<- data.frame()
for (i in 1:ncol(asv.sample)) {
    asv<- data.frame()
    asv[1,1]<- sum(asv.sample[,i]!=0)
    rownames(asv)<- paste0("Sample", i)
    test <- rbind(test, asv) ### Join all the "individual" data frames into the final data frame 
}
asv.sample<- as.matrix(test)
colnames(asv.sample)<- "ASVs_dada2"
rm(test,asv, i)
#Important not to loose the connection between the seq and the ASVs
make_csv_table(asv.names, "/fast/AG_Forslund/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/Tables/asv_table_with_seqs", info = "asv_table_with_seqs")
write.csv(asv.names, "/fast/AG_Forslund/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/Tables/asv_table.csv", row.names = F, col.names = F)

###Track reads through the pipeline 
getN <- function(x) sum(getUniques(x))
track <- cbind(sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(filtFs, getN), sapply(filtRs, getN), sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
#If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
track<-track[,c(1,3,5:8)]
colnames(track) <- c("input_dada2", "filtered_dada2", "denoisedF_dada2", "denoisedR_dada2", "merged_dada2", "nonchim_dada2")
rownames(track) <- sample.names
rownames(asv.sample) <- sample.names
track<-cbind(track, asv.sample)
head(track)

  

taxa <- assignTaxonomy(seqtab.nochim, "/fast/AG_Forslund/Lola/SILVA_databases/SILVA_db", multithread=TRUE)

taxa <- addSpecies(taxa, "/fast/AG_Forslund/Lola/SILVA_databases/silva_species_assignment_v138.1.fa.gz")

taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)

##Create an Taxa matrix with ASV as rows and taxonomic level as columns
taxa_asvs <- as.data.frame(taxa) # Removing sequence rownames for display only
rownames(taxa_asvs) <- NULL
rownames(taxa_asvs) <- paste0("ASV", 1:nrow(asv.names))
head(taxa_asvs)
make_csv_table(taxa_asvs, "/fast/AG_Forslund/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/Tables/tax_table_names", info = "tax_table_n")
make_csv_table(taxa, "/fast/AG_Forslund/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/Tables/tax_table_with_seqs", info = "tax_table_with_seqs")


##Transform seqtab.nochim to fasta file 
  library(DECIPHER); packageVersion("DECIPHER")
  ##Create a DNAString set from the ASVs
  dna <- DNAStringSet(getSequences(seqtab.nochim))
  names(dna)<- paste0("ASV", 1:length(dna)) ##Give short names to each sequence
 writeXStringSet(dna, "/fast/AG_Forslund/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/ASV.fasta") ##Export fasta seq
  
  ##Create a Phylogenetic tree
  library(phangorn)
  Align16S<- AlignSeqs(dna, anchor= NA, verbose= FALSE) ##Alignment
  
  phangAlign16S <- phyDat(as(Align16S, "matrix"), type="DNA")
  dm16S <- dist.ml(phangAlign16S) ## Distance matrix
  treeNJ16S <- NJ(dm16S) # Note, tip order != sequence order
  plot(treeNJ16S, type= "unrooted", use.edge.length= FALSE, no.margin= TRUE, show.tip.label= TRUE) ##Neighbor-Joining tree
  fit1 <- pml(treeNJ16S, data=phangAlign16S)
  fitGTR16S <- update(fit1, k=4, inv=0.2)
  fitGTR16S <- optim.pml(fitGTR16S, model="GTR", optInv=TRUE, optGamma=TRUE,
                         rearrangement = "stochastic", control = pml.control(trace = 0))
  plot(fitGTR16S, type= "unrooted", use.edge.length= FALSE, no.margin= TRUE, show.tip.label= FALSE)
  
  tree<- fitGTR16S$tree

saveRDS(fitGTR16S, "/fast/AG_Forslund/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/phylo_tree.rds")  
saveRDS(seqtab.nochim, "/fast/AG_Forslund/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/seqtab.nochim.ratshiperammon.Rds") 


library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")
theme_set(theme_bw())
samples.out <- rownames(seqtab.nochim)


```