---
title: "4_raref_anal_and_filter_hiperammon"
output: html_document
date: "2023-06-19"
---

#Data exploration
```{r Rarefaction curve}
library("phyloseq"); packageVersion("phyloseq")
library(readr)
library(data.table)
library("ggplot2"); packageVersion("ggplot2")

ps.noncontam <- readRDS("~/Documents/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/ps.noncontam.Rds")
library(readxl)
metadata_hiper <- read_excel("~/Documents/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/metadata_hiperammon.xlsx")
View(metadata_hiper)
ps_filtered1 <-subset_samples(ps.noncontam, SampleID != "Pool_control_S51")
metadata_hiper<-as.data.frame(metadata_hiper)
rownames(metadata_hiper)<-metadata_hiper$SampleID
sample_data(ps_filtered1) <- metadata_hiper

knitr::opts_chunk$set(echo = TRUE)#Rarefaction curve

#Imp not to include the neg control
library(vegan)
S <- specnumber(as.data.frame(t(otu_table(ps_filtered1)))) # observed number of species
(raremax <- min(rowSums(as.data.frame(t(otu_table(ps_filtered1))))))
Srare <- rarefy(t(as.data.frame(otu_table(ps_filtered1))), raremax)
plot(S, Srare, xlab = "Observed No. of Species", ylab = "Rarefied No. of Species")
abline(0, 1)
rarecurve(as.data.frame(t(otu_table(ps_filtered1))), step = 20, sample = raremax, col = "black", cex = 0.6)
saveRDS(ps_filtered1, "~/Documents/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/ps_filtered1.Rds")
```

```{r exploration of the data}
sdt = data.table(as(sample_data(ps_filtered1), "data.frame"),
                 TotalReads = sample_sums(ps_filtered1), keep.rownames = TRUE)
pSeqDepth = ggplot(sdt, aes(TotalReads)) + geom_histogram() + ggtitle("Sequencing Depth")
pSeqDepth
pSeqDepth + facet_wrap(~Group)
pSeqDepth + facet_wrap(~Batch)
```


```{r Removing the one that does not arrive to saturation (=0 all arrive to saturation)}
ps_filtered2<-prune_samples(names(which(sample_sums(ps_filtered1) >= 3000)), ps_filtered1)
ps_filtered2
library(vegan)
S <- specnumber(as.data.frame(t(otu_table(ps_filtered2)))) # observed number of species
(raremax <- min(rowSums(as.data.frame(t(otu_table(ps_filtered2))))))
Srare <- rarefy(t(as.data.frame(otu_table(ps_filtered2))), raremax)
plot(S, Srare, xlab = "Observed No. of Species", ylab = "Rarefied No. of Species")
abline(0, 1)
rarecurve(as.data.frame(t(otu_table(ps_filtered2))), step = 20, sample = raremax, col = "black", cex = 0.6)
saveRDS(ps_filtered2, "~/Documents/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/ps_filtered2_intermediate.Rds")
```

#Seq depth groups wrap
```{r exploration of the data}
ggplot(sdt, aes(Quant_DNA, TotalReads, color = as.factor(Group))) + 
    geom_point(size = 5) + 
    geom_smooth(method = lm) +
    ggtitle("Sequencing Depth vs. DNAquant") +
    scale_y_log10() +
    facet_grid(Group ~ .)

ggplot(sdt, aes(Quant_DNA, TotalReads, color = as.factor(Batch))) + 
    geom_point(size = 5) + 
    geom_smooth(method = lm) +
    ggtitle("Sequencing Depth vs. DNAquant") +
    scale_y_log10() 
```

```{r tree plot}
plot_tree(ps_filtered2, size = "Abundance",
          color = "Group",
          justify = "yes please", 
          ladderize = "left") +
    scale_size_continuous(range = c(1, 3))
```

#Rarefaction1
```{r rarefaction}
# keep result reproductive
 set.seed(321)
 ps.rarefied = rarefy_even_depth(ps_filtered2, rngseed=1, replace=F)
 ps.rarefied
 saveRDS(ps.rarefied, "~/Documents/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/ps.rarefied.Rds")
```

###Exploration
```{r exploration for filteringtaxa}
tdt = data.table(tax_table(ps_filtered2),
                 TotalCounts = taxa_sums(ps_filtered2),
                 OTU = taxa_names(ps_filtered2))
ggplot(tdt, aes(TotalCounts)) + 
    geom_histogram() + 
    ggtitle("Histogram of Total Counts")

# How many singletons?
tdt[(TotalCounts <= 0), .N]
tdt[(TotalCounts <= 1), .N]
# None. This data has been filtered already. 
# How many doubletons?
tdt[(TotalCounts <= 2), .N]

#Filtering Threshold plot
# taxa cumulative sum
taxcumsum = tdt[, .N, by = TotalCounts]
setkey(taxcumsum, TotalCounts)
taxcumsum[, CumSum := cumsum(N)]
# Define the plot
pCumSum = ggplot(taxcumsum, aes(TotalCounts, CumSum)) + 
    geom_point() +
    xlab("Filtering Threshold, Minimum Total Counts") +
    ylab("OTUs Filtered") +
    ggtitle("OTUs that would be filtered vs. the minimum count threshold")
pCumSum
#Zoom in
pCumSum + xlim(0, 100)
```


#2nd taxa filtering: Remove low prevalent taxa
###For AphaDiv (keeping singletons)
It is important to not use filtered data because many richness estimates are modeled on singletons and doubletons in the occurrence table. So, you need to leave them in the dataset if you want a meaningful estimate.
Moreover, we usually not using normalized data because we want to assess the diversity on the raw data and we are not comparing samples to each other but only assessing diversity within each sample.
```{r tree plot}
 # Remove samples with less than MINREADS from phyloseq object
ps_0 <- prune_taxa(taxa_sums(ps_filtered2) > 0, ps_filtered2)
# merge all taxa that are detected rare
total_samples <- phyloseq::nsamples(ps_0)
ps_0
saveRDS(ps_0, "~/Documents/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/ps_0.Rds")
```

###For all the other tests
```{r tree plot}
library(microbiome)
##Create a prevalence dataframe 
Prevdf<- apply(X = otu_table(ps_filtered2),
               MARGIN = 1,
               FUN = function(x){sum(x > 0)})

##Add taxonomy and total read counts to this data.frame
Prevdf<- data.frame(Prevalence = Prevdf,
                    TotalAbundance = taxa_sums(ps_filtered2),
                    tax_table(ps_filtered2))

plyr::ddply(Prevdf, "Phylum", function(df1){
  data.frame(mean_prevalence=mean(df1$Prevalence),total_abundance=sum(df1$TotalAbundance,na.rm = T),stringsAsFactors = F)
})

ggplot(Prevdf, aes(TotalAbundance, Prevalence / nsamples(ps_filtered2),color=Phylum)) +
  geom_hline(yintercept = 0.1, alpha = 2, linetype = 2) + geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Log 10 Total Reads") + ylab("Prevalence [Prop. of Samples]") +
  theme_bw()+
  facet_wrap(~Phylum) + theme(legend.position="none")
ggplot(Prevdf, aes(TotalAbundance, Prevalence / nsamples(ps_filtered2),color=Phylum)) +
    geom_hline(yintercept = 0.1, alpha = 2, linetype = 2) + geom_point(size = 2, alpha = 0.7) +
    scale_x_log10() +  xlab("Log 10 Total Reads") + ylab("Prevalence [Prop. of Samples]") +
    theme_bw()+
    facet_wrap(~Order) + theme(legend.position="right")

#Remove low prevalent ASVs (This data is used for differential abundance analysis)
##Remove ASVs that do not show more than 5 times in more than 10% of the samples
wh0 <- genefilter_sample(ps_filtered2, filterfun_sample(function(x) x > 5), A=0.1*nsamples(ps_filtered1))
ps_filtered2<- prune_taxa(wh0, ps_filtered2)
hist(readcount(ps_filtered2))
ps_filtered2
saveRDS(ps_filtered2, "~/Documents/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/ps_filtered2.Rds")
```

#No need because all the taxa are kept
#Rarefaction2 after filtering
#```{r rarefaction}
# keep result reproductive
# set.seed(321)
# ps.rarefied.filtered = rarefy_even_depth(ps_filtered2, rngseed=1, replace=F)
# ps.rarefied.filtered
# #saveRDS(ps.rarefied.filtered, "~/Documents/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/ps.rarefied.Rds")
#```

#Data transformation for Beta Div
```{r rarefaction}
ps_comp<- microbiome::transform(ps_filtered2, "compositional")
saveRDS(ps_comp, "~/Documents/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/ps_comp.Rds")

ps_clr<- microbiome::transform(ps_filtered2, "clr")
saveRDS(ps_clr, "~/Documents/Lola/Datos_CIPF_2023/Rats_HE_HiperAmmon_model/outputs/ps_clr.Rds")
```